File: app/src/main/java/com/study/fldemo/MainActivity.java
package com.study.fldemo;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.design.widget.TabLayout;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.text.format.Formatter;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.facebook.drawee.backends.pipeline.Fresco;
import com.facebook.drawee.view.SimpleDraweeView;
import com.mob.MobSDK;
import com.study.fldemo.adapter.VpAdapter;
import com.study.fldemo.bean.UserBean;
import com.study.fldemo.dao.DaoManager;
import com.study.fldemo.event.DialogEvent;
import com.study.fldemo.fragment.AndroidFragment;
import com.study.fldemo.fragment.BaseFragment;
import com.study.fldemo.fragment.ExpandFragment;
import com.study.fldemo.fragment.GirlFragment;
import com.study.fldemo.fragment.IosFragment;
import com.study.fldemo.fragment.VideoFragment;
import com.study.fldemo.fragment.WebFragment;
import com.study.fldemo.share.ShareSDKUtils;
import com.study.fldemo.utils.DensityUtil;
import com.study.fldemo.utils.SpUtils;
import com.study.fldemo.view.LazyViewPager;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import cn.sharesdk.framework.Platform;
import cn.sharesdk.framework.ShareSDK;
import cn.sharesdk.tencent.qq.QQ;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {

    private Toolbar toolbar;
    private TabLayout tl;
    private LazyViewPager vp;
    private NavigationView nv;
    private DrawerLayout dl;
    private ActionBarDrawerToggle toggle;
    private SimpleDraweeView sdv;
    private TextView tvName;
    private int choice = -1;
    private AlertDialog dialog;
    private File cacheDir;
    private Handler handler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            if (dialog != null && dialog.isShowing()) {
                dialog.dismiss();
            }
        }
    };
    private TextView tvTip;
    private SpUtils spUtils;
    private Platform platform;
    //数据库管理对象
    private DaoManager daoManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        cacheDir = getApplicationContext().getCacheDir();

        Fresco.initialize(MainActivity.this);
        MobSDK.init(MainActivity.this, "203acaefb6205", "cf155b70894ce61017f26f2ed19af2c3");
        daoManager = DaoManager.getInstance(getApplicationContext());

        setContentView(R.layout.activity_main);
        spUtils = DefineApplication.spUtils;
        initView();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT &&
                Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT_WATCH) {
            int padding = DensityUtil.px2dip(this, 25);
            toolbar.setPadding(0, padding, 0, 0);
            tl.setPadding(0, padding, 0, 0);
        }
        EventBus.getDefault().register(this);


    }

    private long start = 0;
    private long end = 0;

    private void initView() {


        toolbar = (Toolbar) findViewById(R.id.toolbar);
        tl = (TabLayout) findViewById(R.id.tl);
        vp = (LazyViewPager) findViewById(R.id.vp);
        nv = (NavigationView) findViewById(R.id.nv);
        dl = (DrawerLayout) findViewById(R.id.dl);
        View hearView = nv.getHeaderView(0);
        View view = hearView;
        sdv = (SimpleDraweeView) view.findViewById(R.id.sdv);
        tvName = (TextView) view.findViewById(R.id.tv_name);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        toggle = new ActionBarDrawerToggle(this, dl, R.string.app_name, R.string.app_name);
        toggle.syncState();
        dl.addDrawerListener(toggle);

        DisplayMetrics metrics = getResources().getDisplayMetrics();
        int widthPixels = metrics.widthPixels;
        int heightPixels = metrics.heightPixels;
        ViewGroup.LayoutParams lp = nv.getLayoutParams();
        lp.width = (int) (0.6 * widthPixels);
        lp.height = heightPixels;

        List<BaseFragment> lists = new ArrayList<>();
        List<String> titles = new ArrayList<>();
        lists.add(new AndroidFragment());
        titles.add("Android");
        lists.add(new IosFragment());
        titles.add("Ios");
        lists.add(new GirlFragment());
        titles.add("福利");
        lists.add(new WebFragment());
        titles.add("前端");
        lists.add(new ExpandFragment());
        titles.add("拓展资源");
        lists.add(new VideoFragment());
        titles.add("休息视频");

        tl.setupWithViewPager(vp);

        vp.setAdapter(new VpAdapter(getSupportFragmentManager(), lists, titles));

        dl.addDrawerListener(new DrawerLayout.SimpleDrawerListener() {
            @Override
            public void onDrawerClosed(View drawerView) {
                if (choice == -1) {
                    return;
                }
                switch (choice) {
                    case 1:
                        ShareSDKUtils.Login(QQ.NAME, sdv, tvName);
                        break;
                    case 2:
                        clearAppCache(cacheDir, false);
                        break;
                    case 3:
                        Intent intent = new Intent(MainActivity.this, SearchActivity.class);
                        startActivity(intent);
                        break;
                    case 4:
                        ShareSDKUtils.logout(QQ.NAME, sdv);
                        spUtils.clearData();
                        initHearView();
                        initMenuItem();
                        break;
                }
                updateMenuItem();
                choice = -1;
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                updateMenuItem();
            }

        });

        nv.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {

                String string = "";
                switch (item.getItemId()) {
                    case R.id.menu_login:
                        choice = 1;
                        string = "去登陆...";
                        break;
                    case R.id.menu_cache:
                        choice = 2;
                        string = "正在清除...";
                        handler.sendEmptyMessageDelayed(0, 1000);
                        break;
                    case R.id.menu_search:
                        choice = 3;
                        string = "去百度...";
                        break;
                    case R.id.menu_exit:
                        DefineApplication.loginState = false;
                        choice = 4;
                        string = "正在退出...";
                        handler.sendEmptyMessageDelayed(0, 1000);
                        break;
                }

                showDialog(string);

                item.setCheckable(true);
                dl.closeDrawers();
                return true;
            }
        });

        Observable.create(new ObservableOnSubscribe<Platform>() {
            @Override
            public void subscribe(@io.reactivex.annotations.NonNull ObservableEmitter<Platform> e) throws Exception {
                start = System.currentTimeMillis();
                Log.e("TAG", start + "");
                platform = ShareSDK.getPlatform(QQ.NAME);
                e.onNext(platform);
            }
        }).
                subscribeOn(Schedulers.io()).
                observeOn(AndroidSchedulers.mainThread()).
                subscribe(new Consumer<Platform>() {
                    @Override
                    public void accept(@io.reactivex.annotations.NonNull Platform platform) throws Exception {
                        if (platform.isAuthValid()) {
                            end = System.currentTimeMillis();
                            Log.e("TAG", (end - start) * 1.0f / 1000 + "");
                            DefineApplication.loginState = true;
                            initHearView();
                            initMenuItem();
                        }
                        Log.e("TAG", System.currentTimeMillis() - end + "");
                    }
                });


    }

    private void initHearView() {
        UserBean userBean = spUtils.getUserFromSp();
        tvName.setText(userBean.getName());
        String url = userBean.getUrl();
        if (TextUtils.isEmpty(url)) {
            sdv.setActualImageResource(R.drawable.header);
        } else {
            sdv.setImageURI(url);
        }
    }

    private void updateMenuItem() {
        Menu menu = nv.getMenu();
        MenuItem cacheItem = menu.findItem(R.id.menu_cache);
        long size = getFileSize(cacheDir);
        String fileSize = Formatter.formatFileSize(MainActivity.this, size);
        if (size > 0) {
            cacheItem.setTitle("清理缓存  " + fileSize);
        } else {
            cacheItem.setTitle("清理缓存");
        }
    }

    private void clearAppCache(File file, boolean isDelete) {
        if (file.isDirectory()) {
            File[] listFiles = file.listFiles();
            for (int i = 0; i < listFiles.length; i++) {
                clearAppCache(listFiles[i], true);
            }
        }

        if (isDelete) {
            if (file.isFile()) {
                file.delete();
            } else {
                if (file.listFiles().length == 0) {
                    file.delete();
                }
            }
        }
    }

    public long getFileSize(File file) {
        long size = 0;
        if (file.isDirectory()) {
            File[] listFiles = file.listFiles();
            for (int i = 0; i < listFiles.length; i++) {
                File fileItem = listFiles[i];
                size += getFileSize(fileItem);
            }
        } else {
            size += file.length();
        }
        return size;
    }

    @Subscribe
    public void onEventMainThread(DialogEvent event) {
        if (dialog != null && dialog.isShowing()) {
            dialog.dismiss();
        }
        Log.e("thread::", Thread.currentThread().getName() + "==");
        if (event.getState() == 1) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    initMenuItem();
                }
            });

        }
    }

    private void initMenuItem() {
        MenuItem itemLogin = nv.getMenu().findItem(R.id.menu_login);
        MenuItem itemExit = nv.getMenu().findItem(R.id.menu_exit);
        if (platform.isAuthValid()) {
            itemLogin.setVisible(false);
            itemExit.setVisible(true);
        } else {
            itemLogin.setVisible(true);
            itemExit.setVisible(false);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (toggle.onOptionsItemSelected(item)) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    public void showDialog(String string) {
        if (dialog == null) {
            View inflate = View.inflate(this, R.layout.dialog_tip, null);
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setView(inflate);
            tvTip = (TextView) inflate.findViewById(R.id.tv_tip);
            dialog = builder.create();
        }
        tvTip.setText(string);
        dialog.show();
    }
}

File: app/src/main/java/com/study/fldemo/MainActivity.kt
package com.study.fldemo

import android.app.AlertDialog
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Message

import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.bitmap.CircleCrop
import com.bumptech.glide.request.RequestOptions
import com.google.android.material.navigation.NavigationView
import com.google.android.material.tabs.TabLayout

import androidx.drawerlayout.widget.DrawerLayout
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar

import android.text.TextUtils
import android.text.format.Formatter
import android.util.DisplayMetrics
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView


import com.mob.MobSDK
import com.study.fldemo.adapter.VpAdapter
import com.study.fldemo.bean.UserBean
import com.study.fldemo.dao.DaoManager
import com.study.fldemo.event.DialogEvent
import com.study.fldemo.fragment.AndroidFragment
import com.study.fldemo.fragment.BaseFragment
import com.study.fldemo.fragment.ExpandFragment
import com.study.fldemo.fragment.GirlFragment
import com.study.fldemo.fragment.IosFragment
import com.study.fldemo.fragment.VideoFragment
import com.study.fldemo.fragment.WebFragment
import com.study.fldemo.share.ShareSDKUtils
import com.study.fldemo.utils.DensityUtil
import com.study.fldemo.utils.SpUtils
import com.study.fldemo.view.LazyViewPager

import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

import java.io.File
import java.util.ArrayList

import cn.sharesdk.framework.Platform
import cn.sharesdk.framework.ShareSDK
import cn.sharesdk.tencent.qq.QQ
import io.reactivex.Observable
import io.reactivex.ObservableEmitter
import io.reactivex.ObservableOnSubscribe
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.nv_header_layout.*

class MainActivity : AppCompatActivity() {

    private var tvTip: TextView? = null


    private var choice = -1
    private var dialog: AlertDialog? = null
    private val handler = object : Handler() {
        override fun handleMessage(msg: Message) {
            if (dialog != null && dialog!!.isShowing) {
                dialog!!.dismiss()
            }
        }
    }
    private var spUtils: SpUtils? = null
    private var platform: Platform? = null
    //数据库管理对象
    private var daoManager: DaoManager? = null

    private var start: Long = 0
    private var end: Long = 0

    private var toggle: ActionBarDrawerToggle? = null
    private val circleRequestOptions = RequestOptions().transform(CircleCrop())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        MobSDK.init(this@MainActivity, "203acaefb6205", "cf155b70894ce61017f26f2ed19af2c3")
        platform = ShareSDK.getPlatform(QQ.NAME)
        daoManager = DaoManager.getInstance(applicationContext)

        setContentView(R.layout.activity_main)
        spUtils = DefineApplication.spUtils
        initView()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT_WATCH) {
            val padding = DensityUtil.px2dip(this, 25f)
            toolbar.setPadding(0, padding, 0, 0)
            tl.setPadding(0, padding, 0, 0)
        }
        EventBus.getDefault().register(this)


    }


    private fun initView() {


        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        toggle = ActionBarDrawerToggle(this, dl, R.string.app_name, R.string.app_name)
        toggle?.syncState()
        toggle?.apply {
            dl.addDrawerListener(this)
        }

        val metrics = resources.displayMetrics
        val widthPixels = metrics.widthPixels
        val lp = nv!!.layoutParams
        lp.width = (0.6 * widthPixels).toInt()
        nv!!.layoutParams = lp
        val lists = ArrayList<BaseFragment>()
        val titles = ArrayList<String>()
        lists.add(AndroidFragment())
        titles.add("Android")
        lists.add(IosFragment())
        titles.add("Ios")
        lists.add(GirlFragment())
        titles.add("福利")
        lists.add(WebFragment())
        titles.add("前端")
        lists.add(ExpandFragment())
        titles.add("拓展资源")
        lists.add(VideoFragment())
        titles.add("休息视频")

        tl!!.setupWithViewPager(vp)

        vp!!.adapter = VpAdapter(supportFragmentManager, lists, titles)

        dl!!.addDrawerListener(object : DrawerLayout.SimpleDrawerListener() {
            override fun onDrawerClosed(drawerView: View) {
                if (choice == -1) {
                    return
                }
                when (choice) {
                    1 -> ShareSDKUtils.Login(QQ.NAME, sdv, tv_name)
                    2 -> clearAppCache(cacheDir!!, false)
                    3 -> {
                        val intent = Intent(this@MainActivity, SearchActivity::class.java)
                        startActivity(intent)
                    }
                    4 -> {
                        ShareSDKUtils.logout(QQ.NAME, sdv)
                        spUtils!!.clearData()
                        initHearView()
                        initMenuItem()
                    }
                    5 -> {
                        val collectIntent = Intent(this@MainActivity, CollectActivity::class.java)
                        startActivity(collectIntent)
                    }
                }
                updateMenuItem()
                choice = -1
            }

            override fun onDrawerOpened(drawerView: View) {
                super.onDrawerOpened(drawerView)
                updateMenuItem()
            }

        })

        nv!!.setNavigationItemSelectedListener { item ->
            var string = ""
            when (item.itemId) {
                R.id.menu_login -> {
                    choice = 1
                    string = "去登陆..."
                }
                R.id.menu_cache -> {
                    choice = 2
                    string = "正在清除..."
                    handler.sendEmptyMessageDelayed(0, 1000)
                }
                R.id.menu_search -> {
                    choice = 3
                    string = "去百度..."
                }
                R.id.menu_exit -> {
                    DefineApplication.loginState = false
                    choice = 4
                    string = "正在退出..."
                    handler.sendEmptyMessageDelayed(0, 1000)
                }
                R.id.menu_collect -> {
                    choice = 5
                    string = "前往收藏..."
                    handler.sendEmptyMessageDelayed(0, 1000)
                }
            }

            showDialog(string)

            item.isCheckable = true
            dl!!.closeDrawers()
            true
        }

        Observable.create(ObservableOnSubscribe<Platform> { e ->
            start = System.currentTimeMillis()
            Log.e("TAG", start.toString() + "")

            platform?.apply { e.onNext(this) }
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe { platform ->
            if (platform.isAuthValid) {
                end = System.currentTimeMillis()
                Log.e("TAG", ((end - start) * 1.0f / 1000).toString() + "")
                DefineApplication.loginState = true
                initMenuItem()
            }
            initHearView()
            Log.e("TAG", (System.currentTimeMillis() - end).toString() + "")
        }


    }


    private fun initHearView() {
        val userBean = spUtils!!.userFromSp
        tv_name.text = userBean.name
        val url = userBean.url
        if (TextUtils.isEmpty(url)) {
            Glide.with(applicationContext).load(R.drawable.header).apply(circleRequestOptions).into(sdv!!)
        } else {
            Glide.with(applicationContext).load(url).apply(circleRequestOptions).into(sdv!!)
        }
    }

    private fun updateMenuItem() {
        val menu = nv.menu
        val cacheItem = menu.findItem(R.id.menu_cache)
        val size = getFileSize(applicationContext.cacheDir)
        val fileSize = Formatter.formatFileSize(this@MainActivity, size)
        if (size > 0) {
            cacheItem.title = "清理缓存  $fileSize"
        } else {
            cacheItem.title = "清理缓存"
        }
    }

    private fun clearAppCache(file: File, isDelete: Boolean) {
        if (file.isDirectory) {
            val listFiles = file.listFiles()
            for (i in listFiles!!.indices) {
                clearAppCache(listFiles[i], true)
            }
        }

        if (isDelete) {
            if (file.isFile) {
                file.delete()
            } else {
                if (file.listFiles()?.isEmpty() == true) {
                    file.delete()
                }
            }
        }
    }

    private fun getFileSize(file: File): Long {
        var size: Long = 0
        if (file.isDirectory) {
            val listFiles = file.listFiles()
            for (i in listFiles!!.indices) {
                val fileItem = listFiles[i]
                size += getFileSize(fileItem)
            }
        } else {
            size += file.length()
        }
        return size
    }

    @Subscribe
    fun onEventMainThread(event: DialogEvent) {
        if (dialog != null && dialog!!.isShowing) {
            dialog!!.dismiss()
        }

        if (event.state == 1) {
            runOnUiThread { initMenuItem() }

        }
    }

    private fun initMenuItem() {
        val itemLogin = nv!!.menu.findItem(R.id.menu_login)
        val itemExit = nv!!.menu.findItem(R.id.menu_exit)
        if (platform!!.isAuthValid) {
            itemLogin.isVisible = false
            itemExit.isVisible = true
        } else {
            itemLogin.isVisible = true
            itemExit.isVisible = false
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (toggle!!.onOptionsItemSelected(item)) {
            true
        } else super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        EventBus.getDefault().unregister(this)
    }


    private fun showDialog(string: String) {
        if (dialog == null) {
            val inflate = View.inflate(this, R.layout.dialog_tip, null)
            val builder = AlertDialog.Builder(this)
            builder.setView(inflate)
            tvTip = inflate.findViewById<View>(R.id.tv_tip) as TextView
            dialog = builder.create()
        }
        tvTip?.text = string
        dialog?.show()
    }
}
